### PR TITLE
chore(ux): better handle uploading of workflow

### LINKED ISSUE #1885: [➕ Feature]: Improve workflow validation
1. Today "id" is being used as an name. This behavior should be deprecated. If an id is supplied, use it, and use name as name https://github.com/keephq/keep/blob/main/keep/parser/parser.py#L37
2. Deleting an workflow and creating another workflow with the same name shows again all old data since deleting a workflow make it is_deleted=1 and not really deletes the workflow (which is ok). Uploading a workflow without an id should generated different uuid.
3. Improve validation - mark workflow with some flag that says it won't run (e.g. configuration error) until something will fix it

### PR DESCRIPTION
**Pull request by rajeshj11**
…or at card level

Closes #1885 
/claim #1885

## 📑 Description

Enhance workflow validators

 

- [x] Today "id" is being used as an name. This behavior should be deprecated. If an id is supplied, use it, and use name as name https://github.com/keephq/keep/blob/main/keep/parser/parser.py#L37

  

- [x]  Deleting an workflow and creating another workflow with the same name shows again all old data since deleting a workflow make it is_deleted=1 and not really deletes the workflow (which is ok). Uploading a workflow without an id should generated different uuid.

- [x] Improve validation - mark workflow with some flag that says it won't run (e.g. configuration error) until something will fix it

## ✅ Checks

- [X] My pull request adheres to the code style of this project
- [ ] My code requires changes to the documentation
- [ ] I have updated the documentation as required
- [x] All the tests have passed

## ℹ Additional Information

demo video
https://www.loom.com/share/9f234685255a4d79b6dc3adf279a8437

### CODE REVIEW

**Comment by rajeshj11** [comment]
@shahargl please review

File: keep-ui/app/workflows/[workflow_id]/side-nav-bar.tsx [hunk 1]
```diff
@@ -28,7 +28,7 @@ export default function SideNavBar({ workflow }: { workflow: Workflow }) {
     ];
 
     return (
-        <div className="flex flex-col gap-10 pt-6 top-20 p-1 max-w-[270px]">
```
Comment (shahargl) [inline keep-ui/app/workflows/[workflow_id]/side-nav-bar.tsx:31]: why?

File: keep-ui/app/workflows/[workflow_id]/executions.tsx [hunk 2]
```diff
@@ -153,7 +153,7 @@ export default function WorkflowDetailPage({
     const workflow = { last_executions: data.items } as Partial<Workflow>
     return (
         <>
-            <Card className="relative flex p-4 w-full gap-3">
+            <Card className="relative grid p-4 w-full gap-3 grid-cols-[1fr_4fr]">
```
Comment (shahargl) [inline keep-ui/app/workflows/[workflow_id]/executions.tsx:156]: why?

File: keep/api/core/db.py [hunk 3]
```diff
@@ -276,22 +276,26 @@ def add_or_update_workflow(
     description,
     created_by,
     interval,
-    workflow_raw,
+    update_raw,
     is_disabled,
     updated_by=None,
+    workflow_id=None,
```
Comment (shahargl) [inline keep/api/core/db.py:282]: how can it be None and than you filter by it?

File: keep/parser/parser.py [hunk 4]
```diff
@@ -34,18 +34,14 @@ def _get_workflow_id(self, tenant_id, workflow: dict) -> str:
...
         # for backward compatibility reasons, the id on the YAML is actually the name
         # and the id is a unique generated id stored in the db
-        workflow_name = workflow.get("id")
-        if workflow_name is None:
-            raise ValueError("Workflow dict must have an id")
+        
 
         # get the workflow id from the database
-        workflow_id = get_workflow_id(tenant_id, workflow_name)
+        workflow_id = get_workflow_id(tenant_id, workflow.get("id"))
```
Comment (shahargl) [inline keep/parser/parser.py:40]: and if no id??

File: keep/parser/parser.py [hunk 5]
```diff
@@ -34,18 +34,14 @@ def _get_workflow_id(self, tenant_id, workflow: dict) -> str:
...
+        
 
         # get the workflow id from the database
-        workflow_id = get_workflow_id(tenant_id, workflow_name)
+        workflow_id = get_workflow_id(tenant_id, workflow.get("id")) 
         # if the workflow id is not found, it means that the workflow is not stored in the db
         # for example when running from CLI
         # so for backward compatibility, we will use the workflow name as the id
         # todo - refactor CLI to use db also
-        if not workflow_id:
```
Comment (shahargl) [inline keep/parser/parser.py:47]: ????

File: keep/workflowmanager/workflowstore.py [hunk 6]
```diff
@@ -37,16 +37,29 @@ def get_workflow_execution(self, tenant_id: str, workflow_execution_id: str):
...
+            return True 
+
+    
     def create_workflow(self, tenant_id: str, created_by, workflow: dict):
         workflow_id = workflow.get("id")
-        self.logger.info(f"Creating workflow {workflow_id}")
         interval = self.parser.parse_interval(workflow)
         if not workflow.get("name"):  # workflow name is None or empty string
-            workflow_name = workflow_id
+            workflow_name = workflow_id if self._is_not_uuid(workflow_id) else "[No Workflow Name]"
```
Comment (shahargl) [inline keep/workflowmanager/workflowstore.py:53]: workflow must have a name

**Comment by shahargl** [comment]
@rajeshj11 hey, left few comments. IMO its not ready.

also - please add tests so this change will be testable. the best way is to think of the tests, make sure they are failing before your change, and show that your changes make them pass.

**Review by shahargl — CHANGES_REQUESTED** [review]
see comments

File: keep-ui/app/workflows/[workflow_id]/side-nav-bar.tsx (hunk 1 above)
Comment (rajeshj11) [reply keep-ui/app/workflows/[workflow_id]/side-nav-bar.tsx:31]: In this issue context, it is irrelevant but This style change gives us more responsiveness than before.

Note: I only created this component. I have tested it.

File: keep-ui/app/workflows/[workflow_id]/executions.tsx (hunk 2 above)
Comment (rajeshj11) [reply keep-ui/app/workflows/[workflow_id]/executions.tsx:156]: Same as above comment 
https://github.com/keephq/keep/pull/1907#discussion_r1759767095

**Comment by rajeshj11** [comment]
> @rajeshj11 hey, left few comments. IMO its not ready.
> 
> also - please add tests so this change will be testable. the best way is to think of the tests, make sure they are failing before your change, and show that your changes make them pass.

ack

**Comment by rajeshj11** [comment]
@shahargl Just want to understand a few things.

1. how do we provision workflows? is there any command for it? (is it via CLI only or through WEB also)
2. can we update workflows using CLI(other than provision workflows)

**Comment by rajeshj11** [comment]
> @shahargl Just want to understand a few things.
> 
> 1. how do we provision workflows? is there any command for it? (is it via CLI only or through WEB also)
> 2. can we update workflows using CLI(other than provision workflows)

@talboren ^^

**Comment by rajeshj11** [comment]
@shahargl please review

**Comment by shahargl** [comment]
@rajeshj11 will review tomorrow, sorry had a busy day

**Comment by rajeshj11** [comment]
https://www.loom.com/share/9f234685255a4d79b6dc3adf279a8437
demo video^^

File: keep/workflowmanager/workflowstore.py [hunk 7]
```diff
@@ -38,26 +38,49 @@ def get_workflow_execution(self, tenant_id: str, workflow_execution_id: str):
...
+            return True 
+
+    
     def create_workflow(self, tenant_id: str, created_by, workflow: dict):
         workflow_id = workflow.get("id")
-        self.logger.info(f"Creating workflow {workflow_id}")
         interval = self.parser.parse_interval(workflow)
         if not workflow.get("name"):  # workflow name is None or empty string
-            workflow_name = workflow_id
+            workflow_name = workflow_id if workflow_id and self._is_not_uuid(workflow_id) else "[No Workflow Name]"
```
Comment (shahargl) [inline keep/workflowmanager/workflowstore.py:54]: workflow cannot be created without name

File: keep/api/core/db.py [hunk 8]
```diff
@@ -306,24 +306,42 @@ def add_or_update_workflow(
     description,
     created_by,
     interval,
-    workflow_raw,
+    update_raw,
```
Comment (shahargl) [inline keep/api/core/db.py:309]: why did you change it?

File: keep/api/core/db.py [hunk 9]
```diff
@@ -306,24 +306,42 @@ def add_or_update_workflow(
...
     is_disabled,
+    workflow_id,
     provisioned=False,
     provisioned_file=None,
+    re_provision=False,
```
Comment (shahargl) [inline keep/api/core/db.py:314]: what's re_provision?

File: keep/api/core/db.py [hunk 10]
```diff
@@ -306,24 +306,42 @@ def add_or_update_workflow(
...
+        else : 
+            existing_workflow = (
+                session.query(Workflow)
+                .filter_by(id=workflow_id)
+                .filter_by(tenant_id=tenant_id)
+                .first()
+            )
 
         if existing_workflow:
+            workflow_raw = update_raw(existing_workflow.id)
```
Comment (shahargl) [inline keep/api/core/db.py:340]: ??????

**Review by shahargl — CHANGES_REQUESTED** [review]
1. please split the unrelated changes to a different PR it makes hard to understand whats happening
2. I don't understand what you did with workflow_name and workflow_id
3. I don't understand whats re_provision

File: keep/api/core/db.py [hunk 11]
```diff
@@ -306,24 +306,42 @@ def add_or_update_workflow(
     description,
     created_by,
     interval,
-    workflow_raw,
+    update_raw,
     is_disabled,
+    workflow_id,
```
Comment (shahargl) [inline keep/api/core/db.py:311]: but it already get id, so what's workflow_id?

**Comment by Matvey-Kuk** [comment]
@rajeshj11 I appreciate your contribution! Unfortunately I see there are a lot of pieces hard for understand for me and for @shahargl . I will close the PR for now, please open another PR (or PRs) once it's ready for the review.

**Comment by rajeshj11** [comment]
> @rajeshj11 I appreciate your contribution! Unfortunately I see there are a lot of pieces hard for understand for me and for @shahargl . I will close the PR for now, please open another PR (or PRs) once it's ready for the review.

I had previously requested a 1:1 call with @shahargl but didn’t receive a response. I respect your decision to close the PR and appreciate your feedback. However, after considering the changes, I believe the approach will remain consistent, even with additional adjustments. Please feel free to unassign me from this task if needed.
